File: BbQ.ChatWidgets.Sample.Shared/WeatherWidget.cs
using BbQ.ChatWidgets.Models;

namespace BbQ.ChatWidgets.Sample.Shared;

/// <summary>
/// A custom Weather widget demonstrating SSE integration for real-time server-pushed weather data.
/// Displays weather information updated via Server-Sent Events.
/// </summary>
public sealed record WeatherWidget(
    string Label,
    string Action,
    string City,           // City name for weather display
    string StreamId        // SSE stream identifier for server-pushed updates
) : ChatWidget(Label, Action)
{
    public override string Purpose =>
        """
        ***Weather Widget***
        Format: <widget>{"type":"weather","label":"Weather","action":"on_weather_update","city":"San Francisco","streamId":"weather-stream-1"}</widget>
        Displays weather data via SSE (Server-Sent Events) for real-time updates.
        Use when you need to show live weather conditions that update from the server.
        """;
}

File: Sample/BbQ.ChatWidgets.Sample.Angular/ClientApp/src/app/widgets/WeatherWidget.ts
import { ChatWidget } from '@bbq-chat/widgets';

/**
 * Client-side representation of the server weather widget.
 * Supports SSE-driven weather updates with configurable stream ID and city.
 */
export class WeatherWidget extends ChatWidget {
  public readonly city?: string;
  public readonly streamId?: string;

  constructor(label: string, action: string, city?: string, streamId?: string) {
    super('weather', label, action);
    this.city = city;
    this.streamId = streamId;
  }

  toObject(): any {
    const base = this as any;
    return {
      type: 'weather',
      label: base.label,
      action: base.action,
      city: this.city,
      streamId: this.streamId,
    };
  }
}
